### ISSUE TITLE
Could not reproduce your LFW-closed set results

### ISSUE THREAD

**Issue by kizilkanat** [AlfredXiangWu/face_verification_experiment#25]
Hi, I want to reproduce your results mentioned in your paper  arXiv:1511.02683  "A Lightened CNN for Deep Face Representation". 
Although you claim that the fc1 layer dimension is 256 in your paper, it is actually 512 in your prototxt file. I used Dlib to detect and extract faces of LFW-closedset. Dlib in addition, rotates faces to align wrt the centers of eye locations horizontally. Then I resized the faces to 128x128 pixels. Then I used your LightenedCNN_A caffemodel release to extract features from fc1 layer, which has 512 dimensionality, using matcaffe latest release. One thing to note is, you did not release the mean binaryproto file of your training set. Don't you apply mean normalization before testing?
I compared the probe features with gallery features using cosine similarity. I obtained very low Rank-1 accuracy (less than %10). What would be the reason of the difference with your results?

**Comment by AlfredXiangWu** [comment]
1. About the feature dimension, please see issues https://github.com/AlfredXiangWu/face_verification_experiment/issues/3
2. The alignment is not only rotating the eyes horizontally, but also setting ec_mc_y and ec_y which is described in paper and readme. As far as I am concerned, different alignment method may lead to the accuracy dropped. 
3. I don't use mean file for training and testing.
4. Do you use the features I provided to obtain the rank-1 evaluation? And Do you obtain the protocol of rank-1 evaluation  from http://biometrics.cse.msu.edu/pubs/databases.html?

**Comment by kizilkanat** [comment]
1. OK, I didnt know that you used eltwise6 for feature extraction, I will try the same. But my Caffe version is newer than yours, I receive errors when I try with your caffe_ftr.py file.
2. To perform the same alignment with you, would you share the file with 5 landmark points of LFW dataset?
3. OK.
4. Yes, I know the rank-1 protocol.

**Comment by AlfredXiangWu** [comment]
I am sorry I could not share the landmarks of LFW. 
You can use Yi Sun's work(http://mmlab.ie.cuhk.edu.hk/archive/CNN_FacePoint.htm) instead. The images which the points are not detected may be fixed manually.

**Comment by kizilkanat** [comment]
We are now able to extract features from eltwise6 layer of LightenedCNN_A model, and use it without 2D alignment. But LightenedCNN_B model has no eltwise6 layer. It has eltwise5 and eltwise5_fc1 layers. Which one do you use to extract features from LightenedCNN_B model?
Later, I will give the lfw closed-set rank-1 results after performing the same 2D alignment with you.

**Comment by AlfredXiangWu** [comment]
For B model, the "eltwise_fc1" is the feature layer.

**Comment by kizilkanat** [comment]
By the way, which version of Caffe did you use for training (and testing)? Could you point to a direct download link? I think we have different version than yours.

**Comment by AlfredXiangWu** [comment]
I use the caffe-rc to train and test for the released model. 
https://github.com/BVLC/caffe/releases/tag/rc

**Comment by kizilkanat** [comment]
Can you also share your train_val.prototxt or post its contents here?

**Comment by AlfredXiangWu** [comment]
layers {
  name: "data"
  type:DATA
  top: "data"
  top: "label"
  data_param{
      source: "/home/wuxiang/CASIA_WebFace_align_3_train_lmdb"
      backend: LMDB
      batch_size: 32
    }
  transform_param {
    scale: 0.00390625
    crop_size: 128
    mirror: true
  }
  include: { phase: TRAIN }
}

layers {
  name: "data"
  type: DATA
  top: "data"
  top: "label"
  data_param{
      source: "/home/wuxiang/CASIA_WebFace_align_3_val_lmdb"
      backend: LMDB
      batch_size: 2
    }
  transform_param {
    scale: 0.00390625
    crop_size: 128
    mirror: false
  }
  include: { phase: TEST }
}

The data layer configuration is above and the other details are the same as the deploy.prototxt

**Comment by cheer37** [comment]
why did you set the batch_size to 2 in Test mode?

**Comment by AlfredXiangWu** [comment]
@cheer37 The setting is only to save the GPU memory. Testing net doesn't make sense for training. It is only a signal to observe the network convergence.

**Comment by cheer37** [comment]
ok, I see.

**Comment by stgzr** [comment]
Hi, is there any strategy to deal with unbalanced dataset, like CASIA-WebFace?

**Comment by AlfredXiangWu** [comment]
@stgzr  
In my opinion, it is not necessary to deal with the imbalance data if you only train the CNN by softmax loss. 
If you employ metric learning loss such as triplet or verification, the hard negative or semi-hard negative pair sampling is necessary.

**Comment by stgzr** [comment]
@AlfredXiangWu  
I see, thank you!
Now I am trying to train the LightenedCNN_A model from scratch with CASIA-WebFace(clean version). I wonder how long does it take to converge just as your [solver](https://github.com/AlfredXiangWu/face_verification_experiment/issues/68) setting (under one GPU Tesla K40c)? Is there any tricks or tips during training? Or I just set solver as you did and wait for learning converge?

**Comment by AlfredXiangWu** [comment]
@stgzr 
I recommend you to train LightenedCNN_B directly since model B has better generalization than model A. I trained the model for about 1~2 weeks on GTX 980. The performance is easily reproduced following the setting I mentioned in the paper.
